### PR TITLE
Add `dynamic-codecs` package

### PR DESCRIPTION
**Pull request by lorisleiva**
This PR adds a new `dynamic-codecs` package that allows us to create `Codec` instances on demand (as defined in `@solana/codecs`) from Codama nodes.

This package is a bit low-level since the intention is for it to be used inside another `dynamic-parsers` package which will be created shortly after this PR is merged. The goal of the `dynamic-codecs` package is to focus on the foundations of creating codecs for any nodes.

Namely, it offers a `getNodeCodec` function that, given the full `NodePath` of a node, returns a `Codec<unknown>` that enables encoding and decoding data for that node. Here's an example:

```ts
const codec = getNodeCodec([root, program, definedType]);
const bytes = codec.encode(someData);
const decodedData = codec.decode(bytes);
```

See the [README of the new package](https://github.com/codama-idl/codama/blob/d0fdc216e6b87ed363c9763ec0d43651b4554f0b/packages/dynamic-codecs/README.md) for more information.

### CODE REVIEW

**Comment by lorisleiva** [comment]
* **#289** <a href="https://app.graphite.dev/github/pr/codama-idl/codama/289?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @lorisleiva and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/dynamic-codecs/.eslintrc.cjs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+    extends: ['../../.eslintrc.js'],
+    rules: {
+        'sort-keys-fix/sort-keys-fix': 'off',
```
Comment (lorisleiva) [inline packages/dynamic-codecs/.eslintrc.cjs:4]: Note to self: this rule can stay for this package (bad copy pasta).

File: packages/dynamic-codecs/package.json [hunk 2]
```diff
@@ -0,0 +1,70 @@
+{
+    "name": "@codama/dynamic-codecs",
+    "version": "1.0.0",
+    "description": "Node specifications and helpers for the Codama standard",
```
Comment (lorisleiva) [inline packages/dynamic-codecs/package.json:4]: This needs updating.

File: packages/dynamic-codecs/package.json [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        "test:node": "zx ../../node_modules/@codama/internals/scripts/test-unit.mjs node",
+        "test:react-native": "zx ../../node_modules/@codama/internals/scripts/test-unit.mjs react-native",
+        "test:treeshakability": "zx ../../node_modules/@codama/internals/scripts/test-treeshakability.mjs",
+        "test:types": "zx ../../node_modules/@codama/internals/scripts/test-types.mjs"
+    },
+    "dependencies": {
+        "@codama/errors": "workspace:*",
+        "@codama/nodes": "workspace:*",
+        "@codama/visitors-core": "workspace:*",
+        "@solana/addresses": "2.0.0-rc.3",
```
Comment (lorisleiva) [inline packages/dynamic-codecs/package.json:55]: Let's try and remove this dependency by using a string codec directly.

**Review by lorisleiva — COMMENTED** [review]
Quick self-review

File: packages/dynamic-codecs/test/codecs/SentinelTypeNode.test.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+import { expect, test } from 'vitest';
+
+import { getNodeCodec } from '../../src';
+import { hex } from '../_setup';
+
+test('it encodes sentinel types', () => {
+    const sentinel = constantValueNodeFromBytes('base16', 'ffff');
+    const codec = getNodeCodec([sentinelTypeNode(stringTypeNode('utf8'), sentinel)]);
+    expect(codec.encode('Hello World!')).toStrictEqual(hex('48656c6c6f20576f726c6421ffff'));
+    expect(codec.decode(hex('48656c6c6f20576f726c6421ffff'))).toBe('Hello World!');
```
Comment (ngundotra) [inline packages/dynamic-codecs/test/codecs/SentinelTypeNode.test.ts:11]: Very useful test showing how to use sentinel, IMO

**Review by ngundotra — APPROVED** [review]
LGTM! Looking forward to dynamic-parsers package!

**Pull request merged**